### PR TITLE
Keith help

### PR DESCRIPTION
**Pull request by kmmalpha**

### CODE REVIEW

**Comment by kmmalpha** [comment]
@WeebNetsu

File: src/requests/addresses/delete.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * https://docs.wildduck.email/api/#operation/deleteUserAddress
+ * @param userId ID of the Address
+ * @param addressId ID of the Address
+ */
+export const deleteAnAddress = async (
+	_userId: string,
+	addressId: string,
+): Promise<DefaultResponseModel> => {
+	const url = urlQueryBuilder(`${URL}/${addressId}`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/delete.ts]: No, this URL is going to fail, it is currently pointing to `/addresses/forwarded/{address}` and should be pointing to `/users/{user}/addresses/{address}`, since the URL starts with user, maybe we should move this to the users folder OR we can add another URL constant in the config, I might recommend the former

File: src/requests/addresses/get.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+/**
+ * Request forwarded Addresses information
+ *
+ * https://docs.wildduck.email/api/#operation/getForwardedAddress
+ *
+ * @param addressId ID of the Address
+ */
+export const getForwardedAddressesInformation = async (
+	addressId: string
+): Promise<DefaultResponseModel> => {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: You should build a new response model for this request

File: src/requests/addresses/get.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+/**
+ * Request Addresses information
+ *
+ * https://docs.wildduck.email/api/#operation/getForwardedAddress
+ *
+ * @param userId ID of the User
+ * @param addressId ID of the Address
+ */
+export const requestc = async (
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: requestc? Was this supposed to be `List registered Addresses` (https://docs.wildduck.email/api/#operation/getAddresses)?

File: src/requests/addresses/get.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+ * Get Address info
+ *
+ * https://docs.wildduck.email/api/#operation/resolveAddress
+ *
+ * @param addressId ID of the Address or e-mail address string
+ * @param allowWildcard If true then resolves also wildcard addresses
+ */
+export const getAddressInformation = async (
+	addressId: string,
+	allowWildcard: boolean
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Please take a look at how this is done inside of `getUsers` (`src/requests/users/get.ts`)

File: src/requests/addresses/get.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+): Promise<GetUserResponseModel> => {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Are you sure about this? Even if the two share the exact same items in their schemas, we should create a new model, in case one of them ever changes the information it provides

File: src/requests/addresses/get.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	const url = urlQueryBuilder(`${URL}/${addressId}/updates`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: This URL is incorrect

File: src/requests/addresses/config.ts [hunk 7]
```diff
@@ -0,0 +1 @@
+export const URL = "/addresses/forwarded/{address}";
```
Comment (WeebNetsu) [inline src/requests/addresses/config.ts]: I would make this "/addresses" instead and then specify "/forwarded" and all the other url parameters in the requests

File: src/requests/addresses/get.ts [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+/**
+ * List registered Addresses
+ *
+ * https://docs.wildduck.email/api/#operation/getAddresses
+ *
+ * @param username the users wildduck username
+ */
+export const listRegisteredAddresses = async (
+	username: string
+): Promise<GetUserIdByUsernameResponseModel> => {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Same here, I would rather we create a new model... We can however share sub-models

File: src/requests/addresses/get.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	const url = urlQueryBuilder(`${URL}/resolve/${username}`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: URL is incorrect

File: src/requests/addresses/get.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+/**
+ * List registered Addresses
+ *
+ * https://docs.wildduck.email/api/#operation/getAddresses
+ *
+ * @param username the users wildduck username
+ */
+export const listRegisteredAddresses = async (
+	username: string
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Query parameters are not being provided here, where is username coming from?

**Review by WeebNetsu — CHANGES_REQUESTED** [review]

File: src/requests/addresses/get.ts (hunk 3 above)
Comment (kmmalpha) [reply src/requests/addresses/get.ts]: Typo

File: src/requests/addresses/get.ts (hunk 2 above)
Comment (kmmalpha) [reply src/requests/addresses/get.ts]: I'd like a bit of help with that. A bit shakey.

File: src/requests/addresses/get.ts (hunk 6 above)
Comment (kmmalpha) [reply src/requests/addresses/get.ts]: The video explanation you gave me, did not show me how to deal with URLs so i was using others files so try them.

File: src/requests/addresses/get.ts (hunk 4 above)
Comment (kmmalpha) [reply src/requests/addresses/get.ts]: I'm not entirely sure what it is that I'm missing

File: src/requests/addresses/get.ts (hunk 10 above)
Comment (kmmalpha) [reply src/requests/addresses/get.ts]: Fixed

File: src/requests/addresses/config.ts [hunk 11]
```diff
@@ -0,0 +1 @@
+export const URL = "/addresses/{addressId}";
```
Comment (WeebNetsu) [inline src/requests/addresses/config.ts]: This will not work, take this URL as an example `/addresses/forwarded/{address}` which can be found here (https://docs.wildduck.email/api/#operation/deleteForwardedAddress). Now if we look on line 17 in your src/requests/addresses/delete.ts file, we will see this `${URL}/forwarded/${addressId}`... Now this will generate the following URL:
URL + "/forwarded/" + addressId
Which if we replace the variables with their values will look like this:
"/addresses/{addressId}" + "/forwarded/" + "{addressId}"
Which if added together will look like this:
"/addresses/{addressId}/forwarded/{addressId}"

As you can see, the URL does not look like it should, the URL we are generating later on does not exist.

TL;DR
URL here should just be "/addresses"

File: src/requests/addresses/delete.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * https://docs.wildduck.email/api/#operation/deleteUserAddress
+ * 
+ * @param userId ID of the Address
+ * @param addressId ID of the Address
+ */
+export const deleteAnAddress = async (
+	userId: string,
+	addressId: string,
+): Promise<DeleteAddressModel> => {
```
Comment (WeebNetsu) [inline src/requests/addresses/delete.ts]: Since the response for this endpoint is only "success", we can make the promise return "DefaultResponseModel", no need to create its own response model

File: src/requests/addresses/delete.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+	const url = urlQueryBuilder(`/users/${userId}/forwarded/${addressId}`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/delete.ts]: URL is incorrect

File: src/requests/addresses/delete.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Delete a forwarded Address
+ *
+ * https://docs.wildduck.email/api/#operation/deleteForwardedAddress
+ *
+ * @param addressId ID of the Address
+ */
+export const deleteForwardedAddress = async (
+	addressId: string,
+): Promise<DeleteForwardedAddressModel> => {
```
Comment (WeebNetsu) [inline src/requests/addresses/delete.ts]: No need for its own model, you can just use DefaultResponseModel

File: src/requests/addresses/get.ts [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+
+	const res = await axiosConf.get(url);
+
+	return res.data;
+};
+
+/**
+ * Request Addresses information
+ *
+ * https://docs.wildduck.email/api/#operation/getForwardedAddress
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: This URL is the same one as on line 19

File: src/requests/addresses/get.ts (hunk 4 above)
Comment (WeebNetsu) [reply src/requests/addresses/get.ts]: In the video I sent you, take a look at the part where I explain what to do about **query parameters**

File: src/requests/addresses/get.ts (hunk 6 above)
Comment (WeebNetsu) [reply src/requests/addresses/get.ts]: Selection_143

Take a look at what the url should be (see the linked image)

`${URL}/${addressId}/updates` => URL + "/" + addressId + "/updates" => "/addresses/{address}/updates" 

As you can see, the URLs do not match, in the end the URL should look like "/addresses/resolve/{address}"

File: src/requests/addresses/get.ts [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+	return res.data;
+};
+
+/**
+ * List registered Addresses
+ *
+ * https://docs.wildduck.email/api/#operation/getAddresses
+ *
+ * @param query Partial match of an address
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: No, these are query parameters, check the video I sent you and watch the section on **body parameters** and **query parameters**

File: src/requests/addresses/get.ts [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+	tags: string,
+	requiredTags: string,
+	metaData: boolean,
+	internalData: boolean,
+	limit: number,
+	page: number,
+	next: string,
+	previous: string
+): Promise<ListRegisteredAddressesModel> => {
+	const url = urlQueryBuilder(`${URL}/resolve/${query}`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: https://youtu.be/rfHl-6pVbSE

File: src/requests/addresses/get.ts [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+		access_token: wdData.accessToken,
+	});
+
+	const res = await axiosConf.get(url);
+
+	return res.data;
+};
+/**
+ * List addresses from communication register
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Should be "Lists addresses from email headers. Includes addresses both from sent and received messages. Sorted by last usage."

If wildduck provides an explanation on what an endpoint does, then you can just copy/paste it

File: src/requests/addresses/get.ts [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+ *
+ * https://docs.wildduck.email/api/#operation/getUserAddressregister
+ *
+ * @param userId ID of the User
+ */
+export const listAddressesFromCommunicationRegister = async (
+	userId: string
+): Promise<ListAddressFromCommunicationRegisterResultsModel> => {
+	const url = urlQueryBuilder(`/user/${userId}/restore`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: URL is incorrect

File: src/requests/addresses/get.ts [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+};
+/**
+ * List addresses from communication register
+ *
+ * https://docs.wildduck.email/api/#operation/getUserAddressregister
+ *
+ * @param userId ID of the User
+ */
+export const listAddressesFromCommunicationRegister = async (
+	userId: string
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Missing query parameters

File: src/requests/addresses/get.ts [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+ *
+ * https://docs.wildduck.email/api/#operation/getUserAddresses
+ * 
+ * @param userId ID of the User
+ * @param metaData If true, then includes metaData in the response
+ * @param internalData If true, then includes internalData in the response. Not shown for user-role tokens
+ */
+export const listRegisteredAddressesForUser = async (
+	userId: string,
+	internalData: boolean,
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Please query parameters should be in an object

File: src/requests/addresses/get.ts [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+	metaData?: boolean
+): Promise<ListRegisteredAddressesForUserModel> => {
+	const url = urlQueryBuilder(`/user/${userId}/restore`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: URL is incorrect

File: src/requests/addresses/post.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Create new Address
+ *
+ * https://docs.wildduck.email/api/#operation/createUserAddress
+ *
+ * @param userId ID of the User
+ */
+export const createNewAddress = async (
+	userId: string
```
Comment (WeebNetsu) [inline src/requests/addresses/post.ts]: Missing body parameters

**Review by WeebNetsu — CHANGES_REQUESTED** [review]

File: src/requests/addresses/delete.ts (hunk 12 above)
Comment (kmmalpha) [reply src/requests/addresses/delete.ts]: Ohhh, understood

File: src/requests/addresses/delete.ts [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+
+	const res = await axiosConf.delete(url);
+
+	return res.data;
+};
+
+/**
+ * Delete an Address
+ *
+ * https://docs.wildduck.email/api/#operation/deleteForwardedAddress
```
Comment (WeebNetsu) [inline src/requests/addresses/delete.ts]: Should be "https://docs.wildduck.email/api/#operation/deleteUserAddress"

File: src/requests/addresses/get.ts [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+ *
+ * https://docs.wildduck.email/api/#operation/getUserAddress
+ *
+ * @param userId ID of the User
+ * @param addressId ID of the Address
+ */
+export const getAddressesInfo = async (
+	userId: string,
+	addressId: string,
+	bodyData: UpdateAddressInforBodyParametersModel
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Get requests cannot have body data (only query data)

File: src/requests/addresses/get.ts [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+): Promise<GetAddressesInfoResponseModelModel> => {
+	const url = urlQueryBuilder(`/users/${userId}/${addressId}`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: URL is not correct

File: src/requests/addresses/get.ts [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+		access_token: wdData.accessToken,
+	});
+
+	const res = await axiosConf.put(url, bodyData);
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: put requests should be in the put.ts file, not the get.ts file

File: src/requests/addresses/models.ts [hunk 28]
```diff
@@ -0,0 +1,534 @@
...
+   * Datestring of the time the address was created
+   */
+  "created": string
+}
+
+export interface GetAddressInfoQueryParametersModel {
+  /**
+   * If true then resolves also wildcard addresses
+   */
+  allowWildcard: boolean;
```
Comment (WeebNetsu) [inline src/requests/addresses/models.ts]: Should be optional

File: src/requests/addresses/get.ts [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+export const getAddressInfo = async (
+	addressId: string,
+	queryData: GetAddressInfoQueryParametersModel
+): Promise<GetAddressInfoResponseModel> => {
+	const url = urlQueryBuilder(`${URL}/resolve/${addressId}`, {
+		access_token: wdData.accessToken,
+		...queryData,
+	});
+
+	const res = await axiosConf.put(url);
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Should be a get request

File: src/requests/addresses/get.ts [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+ */
+export const listRegisteredAddresses = async (
+	queryData: ListRegisteredAddressesQueryParametersModel
+): Promise<ListRegisteredAddressesResponseModel> => {
+	const url = urlQueryBuilder(`${URL}`, {
+		access_token: wdData.accessToken,
+		...queryData,
+	});
+
+	const res = await axiosConf.put(url);
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Should be a get request

File: src/requests/addresses/models.ts [hunk 31]
```diff
@@ -0,0 +1,534 @@
...
+   * List of tags associated with the Address
+   */
+  "tags": string[],
+  /**
+   * Datestring of the time the address was created
+   */
+  "created": string
+}
+
+export interface ListRegisteredAddressesQueryParametersModel {
```
Comment (WeebNetsu) [inline src/requests/addresses/models.ts]: You can extend with `PageQueryModel` to decrease number of columns that needs to be specified

File: src/requests/addresses/models.ts [hunk 32]
```diff
@@ -0,0 +1,534 @@
...
+   * If true then the forwarded address is disabled
+   */
+  "forwardedDisabled": boolean,
+  /**
+   * List of forwarding targets
+   */
+  "target"?: string[]
+}
+
+export interface ListRegisteredAddressesResponseModel {
```
Comment (WeebNetsu) [inline src/requests/addresses/models.ts:312]: You can extend with `CursorResponseModel` so you only have to specify the results

File: src/requests/addresses/get.ts [hunk 33]
```diff
@@ -0,0 +1,132 @@
...
+export const listRegisteredAddressesForUser = async (
+	userId: string,
+	queryData: ListRegisteredAddressesForUserQueryParametersModel
+): Promise<ListRegisteredAddressesForUserResponseModel> => {
+	const url = urlQueryBuilder(`/users/${userId}/${URL}`, {
+		access_token: wdData.accessToken,
+		...queryData,
+	});
+
+	const res = await axiosConf.put(url);
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Should be a get request

File: src/requests/addresses/get.ts [hunk 34]
```diff
@@ -0,0 +1,132 @@
...
+ * https://docs.wildduck.email/api/#operation/getUserAddresses
+ *
+ * @param userId ID of the User
+ * @param queryData query parameters for additional options
+ */
+export const listRegisteredAddressesForUser = async (
+	userId: string,
+	queryData: ListRegisteredAddressesForUserQueryParametersModel
+): Promise<ListRegisteredAddressesForUserResponseModel> => {
+	const url = urlQueryBuilder(`/users/${userId}/${URL}`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: No, ${URL} should only be prepended, appending it can cause unexpected issues in the future if ever changed, rather just type "/addresses"

This will also result in the URL turning into this in the future: "/users/dasdsadas//addresses" because URL is "/addresses" not "addresses"

File: src/requests/addresses/get.ts [hunk 35]
```diff
@@ -0,0 +1,132 @@
...
+ * https://docs.wildduck.email/api/#operation/getUserAddressregister
+ *
+ * @param userId ID of the User, Example: 507f1f77bcf86cd799439011
+ * @param queryData query parameters for additional options
+ */
+export const listAddressesFromCommunicationRegister = async (
+	userId: string,
+	queryData: ListAddressesFromCommunicationRegisterQueryParametersModel
+): Promise<ListAddressesFromCommunicationRegisterResponseModel> => {
+	const url = urlQueryBuilder(`/users/${userId}/${URL}`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Do not append ${URL} rather type out "addresses", also this URL should be /users/{user}/addressregister, not /users/{user}/addresses

File: src/requests/addresses/get.ts [hunk 36]
```diff
@@ -0,0 +1,132 @@
...
+		access_token: wdData.accessToken,
+		...queryData,
+	});
+
+	const res = await axiosConf.put(url);
```
Comment (WeebNetsu) [inline src/requests/addresses/get.ts]: Should be a get request

File: src/requests/addresses/post.ts [hunk 37]
```diff
@@ -0,0 +1,60 @@
...
+ * https://docs.wildduck.email/api/#operation/createUserAddress
+ * 
+ * @param userId ID of the User
+ * @param bodyData body parameters to create a user
+ */
+export const createAddress = async (
+	userId: string,
+	bodyData: CreateAddressBodyParameterModel
+): Promise<CreationResponseModel> => {
+	const url = urlQueryBuilder(`/users/${userId}/${URL}/`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/post.ts]: Never append ${URL}

File: src/requests/addresses/post.ts [hunk 38]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * Create new forwarded Address
+ * 
+ * Add a new forwarded email address. Addresses can contain unicode characters
+ * Dots in usernames are normalized so no need to create both "firstlast@example.com" and
+ * "first.last@example.com" Special addresses *@example.com and username@
+ * catches all emails to these domains or users without a registered destination
+ * (requires allowWildcard argument)
+ *
+ * http://docs.wildduck.email/api/#operation/restoreUser
```
Comment (WeebNetsu) [inline src/requests/addresses/post.ts]: http://docs.wildduck.email/api/#operation/createForwardedAddress

File: src/requests/addresses/post.ts [hunk 39]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * @param bodyData body parameters to cancel user deletion
+ */
+export const cancelUserDeletion = async (
```
Comment (WeebNetsu) [inline src/requests/addresses/post.ts]: I am assuming this function was supposed to be something else, body data & function name are incorrect

File: src/requests/addresses/put.ts [hunk 40]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * Update forwarded Address information
+ *
+ * https://docs.wildduck.email/api/#operation/updateForwardedAddress
+ *
+ * @param addressId ID of the Address
+ * @param bodyData data to update on the user
+ */
+export const updateUser = async (
```
Comment (WeebNetsu) [inline src/requests/addresses/put.ts]: Function name is not correct

File: src/requests/addresses/put.ts [hunk 41]
```diff
@@ -0,0 +1,75 @@
...
+ * @param userId ID of the User
+ * @param addressId ID of the Address
+ * @param bodyData data to update on the user
+ */
+export const updateAddressInfo = async (
+	userId: string,
+	addressId: string,
+	bodyData: UpdateAddressInforBodyParametersModel
+): Promise<DefaultResponseModel> => {
+	const url = urlQueryBuilder(`/users/${userId}/${addressId}`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/put.ts]: URL is missing  "/addresses/"

File: src/requests/addresses/models.ts [hunk 42]
```diff
@@ -0,0 +1,534 @@
...
+   * If true then disables forwarded address (stops forwarding messages)
+   */
+  "forwardedDisabled"?: boolean,
+  /**
+   * Autoreply information
+   */
+  "autoreply"?: DateStringAutoreplyModel;
+}
+
+export interface UpdateAddressInforBodyParametersModel {
```
Comment (WeebNetsu) [inline src/requests/addresses/models.ts]: Info > Infor

File: src/requests/addresses/put.ts [hunk 43]
```diff
@@ -0,0 +1,75 @@
...
+		access_token: wdData.accessToken,
+	});
+
+	const res = await axiosConf.put(url, bodyData);
+
+	return res.data;
+};
+
+/**
+ * Rename domain in addresses
```
Comment (WeebNetsu) [inline src/requests/addresses/put.ts:57]: Also add "Renames domain names for addresses, DKIM keys and Domain Aliases" to description

File: src/requests/addresses/put.ts [hunk 44]
```diff
@@ -0,0 +1,75 @@
...
+ * 
+ * https://docs.wildduck.email/api/#operation/renameDomain
+ * 
+ * @param userId ID of the User
+ * @param bodyData data to update on the user
+ */
+export const renameDomainInAddresses = async (
+	userId: string,
```
Comment (WeebNetsu) [inline src/requests/addresses/put.ts]: Does not require userId

File: src/requests/addresses/put.ts [hunk 45]
```diff
@@ -0,0 +1,75 @@
...
+	bodyData: RenameDomainInAddressesBodyParametersModel
+): Promise<DefaultResponseModel> => {
+	const url = urlQueryBuilder(`${URL}/${userId}/renameDomain`, {
```
Comment (WeebNetsu) [inline src/requests/addresses/put.ts]: URL is incorrect

File: src/requests/applicationPassword/config.ts [hunk 46]
```diff
@@ -0,0 +1 @@
+export const URL = "/users/{userId}/asps/{asp}";
```
Comment (WeebNetsu) [inline src/requests/applicationPassword/config.ts]: /{asp} should not be part of base URL

File: src/requests/applicationPassword/delete.ts [hunk 47]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * https://docs.wildduck.email/api/#operation/deleteASP
+ *
+ * @param userId ID of the User
+ * @param asp ID of the Application Password
+ */
+export const deleteAppPassword = async (
+	userId: string,
+	aspId: string
+): Promise<DeleteAppPasswordModel> => {
```
Comment (WeebNetsu) [inline src/requests/applicationPassword/delete.ts]: Can use DefaultResponseModel

File: src/requests/applicationPassword/delete.ts [hunk 48]
```diff
@@ -0,0 +1,26 @@
...
+	const url = urlQueryBuilder(`${URL}/${aspId}`, {
```
Comment (WeebNetsu) [inline src/requests/applicationPassword/delete.ts]: Please look at how variables placeholders inside the URL are handled in `src/requests/submission/post.ts`

**Review by WeebNetsu — CHANGES_REQUESTED** [review]
I haven't gone through every file, but the issues I address are probably in most of the unreviewed files as well, please check all of your files to be correct

File: src/requests/addresses/get.ts (hunk 34 above)
Comment (WeebNetsu) [reply src/requests/addresses/get.ts]: It is now "/users/${userId}/address" but should be "/users/${userId}/addresses"

File: src/requests/applicationPassword/config.ts (hunk 46 above)
Comment (WeebNetsu) [reply src/requests/applicationPassword/config.ts]: When removed remember to update the URLs that are using {asp}

File: src/requests/filters/put.ts [hunk 49]
```diff
@@ -0,0 +1,29 @@
...
+ * https://docs.wildduck.email/api/#operation/updateFilter
+ *
+ * @param userId the users wildduck ID
+ * @param filterId Filters unique ID
+ * @param queryData query parameters for additional options
+ */
+export const updateFilterInformation = async (
+	userId: string,
+	filterId: string,
+	queryData: UpdateFilterInfoBodyParametersModel
```
Comment (WeebNetsu) [inline src/requests/filters/put.ts]: Should be called bodyData, not queryData

File: src/requests/filters/put.ts [hunk 50]
```diff
@@ -0,0 +1,29 @@
...
+): Promise<DefaultResponseModel> => {
+	const url = urlQueryBuilder(`${URL.replace("{userId}", userId).replace("{filterId}", filterId)}`, {
+		access_token: wdData.accessToken,
+		...queryData
```
Comment (WeebNetsu) [inline src/requests/filters/put.ts]: Body data should be passed into the .put() method

**.get and .delete** are the only requests that takes query parameters

File: src/requests/filters/put.ts [hunk 51]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * @param userId the users wildduck ID
+ * @param filterId Filters unique ID
+ * @param queryData query parameters for additional options
+ */
+export const updateFilterInformation = async (
+	userId: string,
+	filterId: string,
+	queryData: UpdateFilterInfoBodyParametersModel
+): Promise<DefaultResponseModel> => {
```
Comment (WeebNetsu) [inline src/requests/filters/put.ts]: Does not only return success, it also returns id

File: src/requests/filters/models.ts [hunk 52]
```diff
@@ -0,0 +1,208 @@
+import { LimitUsageModel } from "../../models/index";
+export interface UpdateFilterInfoBodyParametersModel {}
```
Comment (WeebNetsu) [inline src/requests/filters/models.ts]: Empty?

File: src/requests/filters/models.ts [hunk 53]
```diff
@@ -0,0 +1,208 @@
...
+   */
+  mailbox?: string;
+  /**
+   * An array of forwarding targets. The value could either be an email address or a relay url to next
+   * MX server ("smtp://mx2.zone.eu:25") or an URL where mail contents are POSTed to
+   */
+  targets?: string[];
+}
+
+export interface CreateNewFilterBodyParametersModel {
```
Comment (WeebNetsu) [inline src/requests/filters/models.ts:74]: I saw this exact same structure get used in multiple places, you can make this its own interface, put it inside src/models/index.ts and then just extend the interfaces that uses it

File: src/requests/filters/models.ts [hunk 54]
```diff
@@ -0,0 +1,208 @@
...
+   */
+  ha?: boolean;
+  /**
+   * Message size in bytes. If the value is a positive number then message needs to be larger,
+   * negative then message needs to be smaller than abs(size) value
+   */
+  size?: number;
+}
+
+export interface ActionModel {
```
Comment (WeebNetsu) [inline src/requests/filters/models.ts]: Can potentially be moved to src/models/index.ts

File: src/requests/filters/models.ts [hunk 55]
```diff
@@ -0,0 +1,208 @@
+import { LimitUsageModel } from "../../models/index";
+export interface UpdateFilterInfoBodyParametersModel {}
+
+export interface QueryModel {
```
Comment (WeebNetsu) [inline src/requests/filters/models.ts]: Can potentially be moved to src/models/index.ts

File: src/requests/filters/get.ts [hunk 56]
```diff
@@ -0,0 +1,65 @@
...
+ * https://docs.wildduck.email/api/#operation/getFilter
+ *
+ * @param userId Users unique ID
+ * @param filterId Filters unique ID
+ */
+export const requestFilterInformation = async (
+	userId: string,
+	filterId: string
+): Promise<RequestFilterInfoResponseModel> => {
+	const url = urlQueryBuilder(`${URL.replace("{userId}", userId).replace("{filter}", filterId)}`, {
```
Comment (WeebNetsu) [inline src/requests/filters/get.ts]: {filter} != {filterId}

File: src/requests/filters/models.ts (hunk 55 above)
Comment (WeebNetsu) [reply src/requests/filters/models.ts]: but only if used outside of filter requests

File: src/requests/filters/models.ts (hunk 54 above)
Comment (WeebNetsu) [reply src/requests/filters/models.ts]: but only if used outside of filter requests

File: src/requests/filters/models.ts (hunk 53 above)
Comment (WeebNetsu) [reply src/requests/filters/models.ts:74]: but only if used outside of filter requests, otherwise it can become its own interface but stay inside this file

File: src/requests/filters/models.ts [hunk 57]
```diff
@@ -0,0 +1,208 @@
...
+   * If true, then this filter is ignored
+   */
+  disabled: boolean;
+  /**
+   * List of forwarding targets
+   */
+  targets: [string];
+}
+
+export interface ListAllFiltersResponseModel {
```
Comment (WeebNetsu) [inline src/requests/filters/models.ts]: Hmmm, the structure of this interface looks *very* familiar... Maybe we should extend it with a different interface instead

File: src/requests/filters/models.ts [hunk 58]
```diff
@@ -0,0 +1,208 @@
...
+   * Action to take with a matching message
+   */
+  action: ActionModel;
+  /**
+   * If true, then this filter is ignored
+   */
+  disabled: boolean;
+}
+
+export interface ListAllFiltersQueryParametersModel {
```
Comment (WeebNetsu) [inline src/requests/filters/models.ts]: Hmmm, the structure of this interface looks *very* familiar... Maybe we could extend it with a different interface instead?

File: src/requests/filters/delete.ts [hunk 59]
```diff
@@ -0,0 +1,25 @@
...
+ * https://docs.wildduck.email/api/#operation/deleteFilter
+ *
+ * @param userId Users unique ID
+ * @param filterId Filters unique ID
+ */
+export const deleteFilter = async (
+	userId: string,
+	filterId: string
+): Promise<DefaultResponseModel> => {
+	const url = urlQueryBuilder(`${URL.replace("{userId}", userId).replace("{filter}", filterId)}`, {
```
Comment (WeebNetsu) [inline src/requests/filters/delete.ts]: {filter} != {filterId}

File: src/requests/domainAliases/post.ts [hunk 60]
```diff
@@ -0,0 +1,27 @@
...
+ * 
+ * Create new Domain Alias
+ *
+ * https://docs.wildduck.email/api/#operation/createDomainAlias
+ *
+ * @param bodyData body parameters to reset user password+
+ */
+export const createNewDomainAlias = async (
+	bodyData: CreateNewDomainAliasBodyParameterModel
+): Promise<DefaultResponseModel> => {
```
Comment (WeebNetsu) [inline src/requests/domainAliases/post.ts]: Returns more than just success

File: src/requests/domainAliases/config.ts [hunk 61]
```diff
@@ -0,0 +1 @@
+export const URL = "/domainaliases/{aliasId}";
```
Comment (WeebNetsu) [inline src/requests/domainAliases/config.ts]: should be just `/domainaliases`

File: src/requests/domainAliases/models.ts [hunk 62]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  domain: string;
+  /**
+   * <date-time>
+   * Datestring of the time the alias was created
+   */
+  created: string;
+}
+
+export interface ListRegisteredDomainAliasesQueryParametersModel {
```
Comment (WeebNetsu) [inline src/requests/domainAliases/models.ts:38]: The data in here looks awfully similar to data we can extend (src/models/index.ts) ;)

File: src/requests/domainAliases/models.ts [hunk 63]
```diff
@@ -0,0 +1,99 @@
...
+   * Domain Alias
+   */
+  alias: string;
+  /**
+   * The domain this alias applies to
+   */
+  domain: string;
+}
+
+export interface ListRegisteredDomainAliasesResponseModel {
```
Comment (WeebNetsu) [inline src/requests/domainAliases/models.ts]: The data in here looks awfully similar to data we can extend (src/models/index.ts) ;)

**Review by WeebNetsu — CHANGES_REQUESTED** [review]
